File: week2_string_reverser/solutions/c++/stack_reverser.cpp
#include <iostream>
#include <stack>
#include <string>
#include <cassert>

std::string reverseString(const std::string &str) {
	
	std::stack<char> tmp;
	std::string reversed;


	for(const char& c : str) {
		tmp.push(c);
	}

	while(!tmp.empty()){
		reversed += tmp.top();
		tmp.pop();
	}


	return reversed;
}

// Test cases for reverseString function
void testReverseString() {
    assert(reverseString("hello") == "olleh");
    assert(reverseString("") == "");
    assert(reverseString("a") == "a");
    assert(reverseString("madam") == "madam");
    assert(reverseString("12345") == "54321");
    assert(reverseString("abc123") == "321cba");

    std::cout << "All tests passed!" << std::endl;
}

int main() {
    testReverseString();
    
    std::string input;
    std::cout << "Enter a string to reverse: ";
    std::cin >> input;
    std::cout << "Reversed string: " << reverseString(input) << std::endl;

    return 0;
}

File: week2_string_reverser/solutions/python/stack_reverser.py
# reverse_string.py

import unittest

def reverse_string(s: str) -> str:
    stack = []  # Initialize the stack
    
    # Push all characters onto the stack
    for char in s:
        stack.append(char)
    
    reversed_string = ""  # This will hold the reversed string
    
    # Pop characters from the stack and append to reversed_string
    while stack:
        reversed_string += stack.pop()
    
    return reversed_string

# Unit tests for reverse_string function
class TestReverseString(unittest.TestCase):
    """
    Test case for reverse_string function.
    """

    def test_reverse_hello(self):
        self.assertEqual(reverse_string("hello"), "olleh")

    def test_reverse_empty_string(self):
        self.assertEqual(reverse_string(""), "")

    def test_reverse_single_char(self):
        self.assertEqual(reverse_string("a"), "a")

    def test_reverse_palindrome(self):
        self.assertEqual(reverse_string("madam"), "madam")

    def test_reverse_numbers(self):
        self.assertEqual(reverse_string("12345"), "54321")

    def test_reverse_mixed(self):
        self.assertEqual(reverse_string("abc123"), "321cba")

# Run the tests when the file is executed directly
if __name__ == '__main__':
    unittest.main()
